File: prelude/toolchains/android/src/com/facebook/buck/jvm/cd/command/kotlin/KotlinExtraParams.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under both the MIT license found in the
 * LICENSE-MIT file in the root directory of this source tree and the Apache
 * License, Version 2.0 found in the LICENSE-APACHE file in the root directory
 * of this source tree.
 */

package com.facebook.buck.jvm.cd.command.kotlin;

import com.facebook.buck.core.filesystems.AbsPath;
import com.facebook.buck.core.util.immutables.BuckStyleValue;
import com.facebook.buck.jvm.java.CompileToJarStepFactory;
import com.facebook.buck.jvm.java.ResolvedJavacOptions;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSortedSet;
import java.util.Map;
import java.util.Optional;

/** Extra params for creating Kotlin compile steps. */
@BuckStyleValue
public abstract class KotlinExtraParams implements CompileToJarStepFactory.ExtraParams {

  public static final String KOTLINC_WORKING_DIR = "kotlinc_working_dir";
  public static final String KOTLINC_JVM_ABI_GEN_WORKING_DIR = "jvm_abi_gen_working_dir";

  public abstract ImmutableList<AbsPath> getExtraClassPaths();

  public abstract AbsPath getResolvedStandardLibraryClassPath();

  public abstract AbsPath getResolvedAnnotationProcessingClassPath();

  public abstract AnnotationProcessingTool getAnnotationProcessingTool();

  public abstract ImmutableList<String> getExtraKotlincArguments();

  public LanguageVersion getSanitizedLanguageVersion() {
    String languageVersion = getLanguageVersion();
    if (languageVersion == null) {
      throw new IllegalStateException("Language version is not set");
    }
    return new LanguageVersion(languageVersion);
  }

  protected abstract String getLanguageVersion();

  public abstract ImmutableMap<AbsPath, ImmutableMap<String, String>>
      getResolvedKotlinCompilerPlugins();

  public abstract ImmutableMap<String, AbsPath> getResolvedKosabiPluginOptionPath();

  public abstract Optional<String> getKosabiJvmAbiGenEarlyTerminationMessagePrefix();

  public abstract ImmutableSortedSet<AbsPath> getResolvedFriendPaths();

  public abstract ImmutableList<AbsPath> getResolvedKotlinHomeLibraries();

  public abstract ResolvedJavacOptions getResolvedJavacOptions();

  public abstract Optional<String> getJvmTarget();

  public abstract boolean shouldGenerateAnnotationProcessingStats();

  public abstract boolean shouldUseJvmAbiGen();

  public abstract Optional<AbsPath> getJvmAbiGenPlugin();

  public abstract boolean shouldVerifySourceOnlyAbiConstraints();

  public abstract boolean shouldRemoveKotlinCompilerFromClassPath();

  public abstract boolean shouldKotlincRunViaBuildToolsApi();

  public abstract boolean shouldKotlincRunIncrementally();

  public abstract Optional<AbsPath> getIncrementalStateDir();

  public abstract Optional<AbsPath> getDepTrackerPlugin();

  public abstract boolean shouldUseStandaloneKosabi();

  public Optional<AbsPath> getKotlincWorkingDir() {
    return getIncrementalStateDir().map(dir -> dir.resolve(KOTLINC_WORKING_DIR));
  }

  public Optional<AbsPath> getJvmAbiGenWorkingDir() {
    return getIncrementalStateDir().map(dir -> dir.resolve(KOTLINC_JVM_ABI_GEN_WORKING_DIR));
  }

  /** Package extra params that were resolved before. */
  public static KotlinExtraParams of(
      ImmutableList<AbsPath> extraClassPaths,
      AbsPath standardLibraryClassPath,
      AbsPath annotationProcessingClassPath,
      AnnotationProcessingTool annotationProcessingTool,
      ImmutableList<String> extraKotlincArguments,
      Map<AbsPath, ImmutableMap<String, String>> kotlinCompilerPlugins,
      Map<String, AbsPath> kosabiPluginOptions,
      Optional<String> kosabiJvmAbiGenEarlyTerminationMessagePrefix,
      ImmutableSortedSet<AbsPath> friendPaths,
      ImmutableSortedSet<AbsPath> kotlinHomeLibraries,
      ResolvedJavacOptions resolvedJavacOptions,
      Optional<String> jvmTarget,
      boolean shouldGenerateAnnotationProcessingStats,
      boolean shouldUseJvmAbiGen,
      Optional<AbsPath> jvmAbiGenPlugin,
      boolean shouldVerifySourceOnlyAbiConstraints,
      boolean shouldRemoveKotlinCompilerFromClassPath,
      Optional<AbsPath> depTrackerPlugin,
      boolean shouldKotlincRunViaBuildToolsApi,
      boolean shouldKotlincRunIncrementally,
      boolean shouldUseStandaloneKosabi,
      Optional<AbsPath> incrementalStateDir,
      String languageVersion) {
    return ImmutableKotlinExtraParams.ofImpl(
        extraClassPaths,
        standardLibraryClassPath,
        annotationProcessingClassPath,
        annotationProcessingTool,
        extraKotlincArguments,
        languageVersion,
        kotlinCompilerPlugins,
        kosabiPluginOptions,
        kosabiJvmAbiGenEarlyTerminationMessagePrefix,
        friendPaths,
        kotlinHomeLibraries,
        resolvedJavacOptions,
        jvmTarget,
        shouldGenerateAnnotationProcessingStats,
        shouldUseJvmAbiGen,
        jvmAbiGenPlugin,
        shouldVerifySourceOnlyAbiConstraints,
        shouldRemoveKotlinCompilerFromClassPath,
        shouldKotlincRunViaBuildToolsApi,
        shouldKotlincRunIncrementally,
        incrementalStateDir,
        depTrackerPlugin,
        shouldUseStandaloneKosabi);
  }
}

File: prelude/toolchains/android/src/com/facebook/buck/jvm/cd/command/kotlin/KotlinExtraParams.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is dual-licensed under either the MIT license found in the
 * LICENSE-MIT file in the root directory of this source tree or the Apache
 * License, Version 2.0 found in the LICENSE-APACHE file in the root directory
 * of this source tree. You may select, at your option, one of the
 * above-listed licenses.
 */

package com.facebook.buck.jvm.cd.command.kotlin

import com.facebook.buck.core.filesystems.AbsPath
import com.facebook.buck.jvm.java.CompileToJarStepFactory
import com.facebook.buck.jvm.java.ResolvedJavacOptions
import com.google.common.collect.ImmutableList
import com.google.common.collect.ImmutableMap
import com.google.common.collect.ImmutableSortedSet
import java.util.Optional

data class KotlinExtraParams(
    val extraClassPaths: ImmutableList<AbsPath>,
    val extraClassPathSnapshots: ImmutableList<AbsPath>,
    val standardLibraryClassPath: AbsPath,
    val annotationProcessingClassPath: AbsPath,
    val annotationProcessingTool: AnnotationProcessingTool,
    val extraKotlincArguments: ImmutableList<String>,
    val kotlinCompilerPlugins: ImmutableMap<AbsPath, ImmutableMap<String, String>>,
    val kosabiPluginOptions: ImmutableMap<String, AbsPath>,
    val kosabiJvmAbiGenEarlyTerminationMessagePrefix: Optional<String>,
    val friendPaths: ImmutableSortedSet<AbsPath>,
    val kotlinHomeLibraries: ImmutableList<AbsPath>,
    val resolvedJavacOptions: ResolvedJavacOptions,
    val jvmTarget: Optional<String>,
    val shouldUseJvmAbiGen: Boolean,
    val jvmAbiGenPlugin: Optional<AbsPath>,
    val shouldVerifySourceOnlyAbiConstraints: Boolean,
    val depTrackerPlugin: Optional<AbsPath>,
    val shouldKotlincRunIncrementally: Boolean,
    val incrementalStateDir: Optional<AbsPath>,
    val shouldKsp2RunIncrementally: Boolean,
    private val languageVersionString: String,
    val shouldKosabiJvmAbiGenUseK2: Boolean,
    val kotlinClassesDir: AbsPath,
) : CompileToJarStepFactory.ExtraParams {

  val shouldActionRunIncrementally: Boolean =
      shouldKotlincRunIncrementally || shouldKsp2RunIncrementally

  val kotlincWorkingDir: Optional<AbsPath> =
      incrementalStateDir.map { dir: AbsPath -> dir.resolve(KOTLINC_WORKING_DIR) }

  val ksp2CachesDir: Optional<AbsPath> =
      incrementalStateDir.map { dir: AbsPath -> dir.resolve(KSP2_CACHES_DIR) }

  val jvmAbiGenWorkingDir: Optional<AbsPath> =
      incrementalStateDir.map { dir: AbsPath -> dir.resolve(KOTLINC_JVM_ABI_GEN_WORKING_DIR) }

  val languageVersion: LanguageVersion = LanguageVersion(languageVersionString)

  companion object {
    private val KOTLINC_WORKING_DIR: String = "kotlinc_working_dir"
    private val KSP2_CACHES_DIR: String = "ksp2_caches_dir"
    private val KOTLINC_JVM_ABI_GEN_WORKING_DIR: String = "jvm_abi_gen_working_dir"
  }
}
